src/dbnode/storage/entry.go

```diff
@@ -187,6 +210,44 @@ func (entry *Entry) OnIndexFinalize(blockStartNanos xtime.UnixNano) {
 	entry.DecrementReaderWriterCount()
 }
 
+// IfAlreadyIndexedMarkIndexSuccessAndFinalize marks the entry as successfully
+// indexed if already indexed and returns true. Otherwise returns false.
+func (entry *Entry) IfAlreadyIndexedMarkIndexSuccessAndFinalize(
+	blockStart xtime.UnixNano,
+) bool {
+	successAlready := false
+	entry.reverseIndex.Lock()
+	for _, state := range entry.reverseIndex.states {
+		if state.success {
+			successAlready = true
+			break
+		}
+	}
+	if successAlready {
+		entry.reverseIndex.setSuccessWithWLock(blockStart)
+		entry.reverseIndex.setAttemptWithWLock(blockStart, false)
+	}
+	entry.reverseIndex.Unlock()
+	if successAlready {
+		// indicate the index has released held reference for provided write
+		entry.DecrementReaderWriterCount()
+	}
+	return successAlready
+}
+
+// RelookupAndCheckIsEmpty looks up the series and checks if it is empty.
+// The first result indicates if the series is empty.
+// The second result indicates if the series can be looked up at all.
+func (entry *Entry) RelookupAndCheckIsEmpty() (bool, bool) {
+	e, _, err := entry.Shard.TryRetrieveWritableSeries(entry.Series.ID())
+	if err != nil || e == nil {
+		return false, false
+	}
+	defer entry.DecrementReaderWriterCount()
```

Comment:
As discussed: I believe this should be `defer e.DecrementReaderWriterCount()` to cover the edge case where `entry != e` (i.e. the case where there is multiple competing asynchronous inserts and one entry is orphaned even though the datapoint that was part of the async insert is retained/written).